// user-agent-blocker-buildhooks/edge-functions/ua_blocker_ef.ts
// src/edge-functions/ef.ts
import { getStore } from "@netlify/blobs";

// src/userAgentPolicy.ts
var supportedUserAgents = [
  "GPTBot",
  "GoogleOther",
  "PetalBot",
  "Factset_spyderbot",
  "LINER Bot",
  "ClaudeBot",
  "Timpibot",
  "GoogleBot",
  "BingBot",
  "YandexBot",
  "DuckDuckBot",
  "SemrushBot",
  "AhrefsBot",
  "AhrefsSiteAudit",
  "AwarioBot",
  "DotBot",
  "MJ12bot"
];
var POLICY_GROUP_SIZE = 5;
var NULL_SENTINEL = "<NULL>";
function isAgentBlocked(policy, index) {
  return !!(policy.charCodeAt(1 + Math.floor(index / POLICY_GROUP_SIZE)) & 1 << index % POLICY_GROUP_SIZE);
}
function getAgentsWithNull() {
  return [NULL_SENTINEL, ...supportedUserAgents];
}

// src/edge-functions/ef.ts
var ef_default = (request) => {
  const analyticsEnabled = !!Netlify.env.get("USER_AGENT_BLOCKER_ANALYTICS");
  const policy = Netlify.env.get("USER_AGENT_BLOCKER_EF_POLICY") || "";
  const blockingEnabled = policy[0] === "1";
  if (!analyticsEnabled && !blockingEnabled) {
    return;
  }
  const userAgent = (request.headers.get("user-agent") || "").toLowerCase();
  const willBlock = blockingEnabled && getAgentsWithNull().some((ua, i) => {
    if (ua === NULL_SENTINEL && userAgent.trim() === "") {
      return true;
    } else if (ua !== NULL_SENTINEL && isAgentBlocked(policy, i) && userAgent.includes(ua.toLowerCase())) {
      return true;
    }
  });
  if (analyticsEnabled) {
    if (userAgent.trim() === "") {
      increment(NULL_SENTINEL, willBlock);
    } else {
      for (const ua of supportedUserAgents) {
        if (userAgent.includes(ua.toLowerCase())) {
          increment(ua, willBlock);
          break;
        }
      }
    }
  }
  if (willBlock) {
    console.log(
      `Blocked request ${request.method} ${request.url} from ${userAgent}`
    );
    return new Response("Blocked", {
      status: 403,
      headers: {
        "Content-Type": "text/plain"
      }
    });
  }
};
var config = {
  path: "/*",
  onError: "bypass"
};
async function incrementInBlob(userAgent, willBlock) {
  const store = getStore({
    name: "netlify-user-agent-blocker",
    consistency: "strong"
  });
  const today = (/* @__PURE__ */ new Date()).toISOString().split("T")[0];
  const bucket = Math.floor(Math.random() * 25).toString().padStart(4, "0");
  const key = `counts/${today}/${bucket}`;
  const current = await store.get(key, { type: "json" }) ?? {};
  current[userAgent] ??= { blocked: 0, allowed: 0 };
  const type = willBlock ? "blocked" : "allowed";
  current[userAgent][type] = (current[userAgent][type] || 0) + 1;
  await store.setJSON(key, current);
  const expectedDates = new Set(
    Array.from({ length: 7 }, (_, i) => {
      const date = /* @__PURE__ */ new Date();
      date.setDate(date.getDate() - i);
      return date.toISOString().split("T")[0];
    })
  );
  const { blobs: allCounts } = await store.list({ prefix: "counts/" });
  for (const blob of allCounts) {
    if (!expectedDates.has(blob.key.split("/")[1])) {
      await store.delete(blob.key);
    }
  }
}
function increment(userAgent, willBlock) {
  setTimeout(() => incrementInBlob(userAgent, willBlock), 0);
}
export {
  config,
  ef_default as default
};
