// user-agent-blocker-buildhooks/edge-functions/ua_blocker_robots.ts
// src/userAgentPolicy.ts
var supportedUserAgents = [
  "GPTBot",
  "GoogleOther",
  "PetalBot",
  "Factset_spyderbot",
  "LINER Bot",
  "ClaudeBot",
  "Timpibot",
  "GoogleBot",
  "BingBot",
  "YandexBot",
  "DuckDuckBot",
  "SemrushBot",
  "AhrefsBot",
  "AhrefsSiteAudit",
  "AwarioBot",
  "DotBot",
  "MJ12bot"
];
var POLICY_GROUP_SIZE = 5;
var NULL_SENTINEL = "<NULL>";
function isAgentBlocked(policy, index) {
  return !!(policy.charCodeAt(1 + Math.floor(index / POLICY_GROUP_SIZE)) & 1 << index % POLICY_GROUP_SIZE);
}
function getAgentsWithNull() {
  return [NULL_SENTINEL, ...supportedUserAgents];
}

// src/edge-functions/robots.ts
var robots_default = async (_request, context) => {
  const policy = Netlify.env.get("USER_AGENT_BLOCKER_ROBOTS_POLICY") || "";
  const response = await context.next();
  let text = `User-agent: *
Allow: /
`;
  if (response.status === 200) {
    text = await response.text();
  }
  if (policy[0] === "1") {
    getAgentsWithNull().forEach((ua, i) => {
      if (ua !== NULL_SENTINEL && isAgentBlocked(policy, i)) {
        text += `

User-agent: ${ua}
Disallow: /`;
      }
    });
  }
  return new Response(text, {
    status: 200,
    headers: {
      "Content-Type": "text/plain"
    }
  });
};
var config = {
  path: "/robots.txt"
};
export {
  config,
  robots_default as default
};
